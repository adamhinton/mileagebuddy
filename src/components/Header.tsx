"use client";

// basic Header to go at top of application
// Make sure it fits with styling I've already used in this project
// Should include ThemeSwitch which I've defined elsewhere
// Use dark and light mode functionality

import React from "react";
import ThemeSwitch from "./ThemeSwitch";
import { createClientCSROnly } from "@/app/utils/server/supabase/client";
import { useAppSelector } from "@/redux/hooks";
import Link from "next/link";

export const Header: React.FC = () => {
	const user = useAppSelector((state) => state.user.value);
	console.log("redux user in Header:", user);

	return (
		<header className="bg-white dark:bg-slate-800 p-4 flex justify-between items-center">
			<h1 className="text-2xl bg-white text-black dark:bg-black dark:text-white ">
				Trip Planner
			</h1>
			{/* Just for testing */}
			<h2>{user ? "Logged in" : "Not logged in "}</h2>
			<button
				className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4"
				onClick={async () => {
					const supabase = createClientCSROnly();
					// Also triggers redux store to update via <AuthWatcher/>
					const { error } = await supabase.auth.signOut();
					if (!error) {
						console.log("Logged out");
					} else {
						console.error("Error logging out:", error.message);
					}
				}}
			>
				Sign Out
			</button>
			<nav className="flex space-x-4">
				<Link href="/dashboard">Dashboard</Link>
				<Link href="/login">Login</Link>
				<Link href="/calculator">Calculator</Link>">
			</nav>
			<ThemeSwitch />
		</header>
	);
};
