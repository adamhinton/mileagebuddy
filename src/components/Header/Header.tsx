// ______________________________________________________________________________
// This is (obviously) the main Header.
// // It contains the logo, title, navigation links, dark mode toggle (ThemeSwitch) and auth buttons (login/logout)
// It's split up in to several sub-components
// TODO tabs for internal navigation
// TODO stretch: Put these class names in tailwindClassNames.ts
// ______________________________________________________________________________

"use client";

import React, { useState, useRef, useEffect, RefObject } from "react";
import Link from "next/link";
import ThemeSwitch from "../ThemeSwitch";
import { createClientCSROnly } from "@/app/utils/server/supabase/client";
import { useAppSelector } from "@/redux/hooks";
import ProfileDropdown from "./HeaderSubComponents/ProfileDropdown";
import NavLinks from "./HeaderSubComponents/NavLinks";
import MobileHamburgerMenu from "./HeaderSubComponents/MobileHamburgerMenu";
import Tabs from "./HeaderSubComponents/Tabs";
// TODO header.test

export const Header: React.FC = () => {
	const user = useAppSelector((state) => state.user.value);
	const isLoggedIn = user?.id ? true : false;
	const [isMobileMenuOpen, setIsMobileMenuOpen] = useState(false);
	const [isProfileDropdownOpen, setIsProfileDropdownOpen] = useState(false);
	const profileDropdownRef = useRef<HTMLDivElement>(null);

	const handleSignOut = async () => {
		const supabase = createClientCSROnly();
		// Also triggers redux store to update via <AuthWatcher/>
		const { error } = await supabase.auth.signOut();
		if (error) {
			console.error("Error logging out:", error.message);
		}
	};

	// Close dropdown when clicking outside
	useEffect(() => {
		const handleClickOutside = (event: MouseEvent) => {
			if (
				profileDropdownRef.current &&
				!profileDropdownRef.current.contains(event.target as Node)
			) {
				setIsProfileDropdownOpen(false);
			}
		};

		document.addEventListener("mousedown", handleClickOutside);
		return () => {
			document.removeEventListener("mousedown", handleClickOutside);
		};
	}, []);

	return (
		<header className="bg-background-header border-b border-primary-100 dark:border-primary-200 shadow-sm">
			<div className="container mx-auto px-4 py-3 flex flex-col sm:flex-row sm:items-center sm:justify-between">
				{/* Logo and title section */}
				<div className="flex items-center justify-between">
					<Link href="/" className="flex items-center space-x-2 group">
						<span className="text-2xl font-bold bg-gradient-to-r from-primary via-secondary to-accent bg-clip-text text-transparent">
							MileageBuddy
						</span>
						<span className="hidden md:inline-block text-sm text-neutral-text opacity-70 group-hover:opacity-100 transition-opacity">
							True car cost calculator
						</span>
					</Link>

					{/* Mobile menu hamburger button. Hidden on large screen sizes */}
					<MobileHamburgerMenu
						isMobileMenuOpen={isMobileMenuOpen}
						setIsMobileMenuOpen={setIsMobileMenuOpen}
					/>
				</div>

				{/* Navigation and auth section - visible on larger screens or when mobile menu is open */}
				<div
					className={`${isMobileMenuOpen ? "flex" : "hidden"} sm:flex flex-col sm:flex-row items-center mt-4 sm:mt-0 space-y-3 sm:space-y-0 sm:space-x-6`}
				>
					{isLoggedIn && (
						// Simple Profile dropdown (obviously) that shows user's email and a signout button
						<ProfileDropdown
							user={user!}
							isProfileDropdownOpen={isProfileDropdownOpen}
							setIsProfileDropdownOpen={setIsProfileDropdownOpen}
							handleSignOut={handleSignOut}
							profileDropdownRef={
								profileDropdownRef as unknown as RefObject<HTMLDivElement>
							}
						/>
					)}

					{/* Main navigation*/}
					<NavLinks />

					{/* Dark/light mode toggle icon */}
					<div className="hidden sm:block">
						<ThemeSwitch />
					</div>
				</div>
			</div>

			<Tabs />
		</header>
	);
};
