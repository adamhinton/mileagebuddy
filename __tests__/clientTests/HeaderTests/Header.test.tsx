// _______________________________________________________________
// This is (obviously) tests for Header.tsx
// _______________________________________________________________

import TestReduxStore from "@/app/utils/unitTestUtils/dummyReduxStore";
import { Header } from "@/components/Header/Header";
import { render } from "@testing-library/react";

// TODO hopefully I didn't forget to write these tests
describe("Header.tsx", () => {
	it("Renders without errors", () => {
		render(
			<TestReduxStore>
				<Header />
			</TestReduxStore>
		);
	});
});
