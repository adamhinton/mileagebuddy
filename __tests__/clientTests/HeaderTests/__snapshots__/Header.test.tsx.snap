// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header.tsx Snapshot tests matches snapshot of default state from 4.30.2025 1`] = `
<div>
  <header
    class="bg-background-header border-b border-primary-100 dark:border-primary-200 shadow-sm"
  >
    <div
      class="container mx-auto px-4 py-3 flex flex-col sm:flex-row sm:items-center sm:justify-between"
    >
      <div
        class="flex items-center justify-between"
      >
        <a
          class="flex items-center space-x-2 group"
          href="/"
        >
          <span
            class="text-2xl font-bold bg-gradient-to-r from-primary via-secondary to-accent bg-clip-text text-transparent"
          >
            MileageBuddy
          </span>
          <span
            class="hidden md:inline-block "
          >
            <img
              alt="Project Logo"
              class="h-8 w-8"
              data-nimg="1"
              decoding="async"
              height="8"
              loading="lazy"
              src="/img.jpg"
              style="color: transparent;"
              width="8"
            />
          </span>
        </a>
        <div
          class="sm:hidden flex items-center space-x-3"
        >
          <button
            aria-label="Toggle mobile menu"
            class="p-2 rounded-md text-neutral-text hover:bg-background-highlight"
          >
            <svg
              class="h-6 w-6"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 6h16M4 12h16M4 18h16"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="hidden sm:flex flex-col sm:flex-row items-center mt-4 sm:mt-0 space-y-3 sm:space-y-0 sm:space-x-6"
      >
        <div
          class="relative"
        >
          <button
            class="flex items-center px-3 py-1 bg-background-elevated rounded-full text-sm text-neutral-text border border-primary-50 dark:border-primary-200 hover:bg-background-highlight transition-colors"
          >
            <div
              class="w-2 h-2 rounded-full bg-accent mr-2 animate-pulse"
            />
            <span
              class="hidden sm:inline"
              data-testid="profile-icon"
            >
              Profile
            </span>
            <svg
              class="h-4 w-4 ml-2"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 9l-7 7-7-7"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
        <nav
          class="flex flex-col sm:flex-row items-center space-y-2 sm:space-y-0 space-x-0 sm:space-x-2 w-full sm:w-auto"
        >
          <a
            class="nav-link w-full sm:w-auto text-center"
            href="https://www.linkedin.com/in/adam-hinton/"
            target="_blank"
          >
            Adam Hinton
          </a>
          <a
            class="nav-link w-full sm:w-auto text-center"
            href="/calculator"
          >
            Calculator
          </a>
          <a
            class="nav-link w-full sm:w-auto text-center"
            href="https://github.com/adamhinton/mileagebuddy/"
          >
            Source
          </a>
        </nav>
      </div>
    </div>
  </header>
</div>
`;

exports[`Header.tsx Snapshot tests matches snapshot with profile dropdown open from 4.30.2025 1`] = `
<div>
  <header
    class="bg-background-header border-b border-primary-100 dark:border-primary-200 shadow-sm"
  >
    <div
      class="container mx-auto px-4 py-3 flex flex-col sm:flex-row sm:items-center sm:justify-between"
    >
      <div
        class="flex items-center justify-between"
      >
        <a
          class="flex items-center space-x-2 group"
          href="/"
        >
          <span
            class="text-2xl font-bold bg-gradient-to-r from-primary via-secondary to-accent bg-clip-text text-transparent"
          >
            MileageBuddy
          </span>
          <span
            class="hidden md:inline-block "
          >
            <img
              alt="Project Logo"
              class="h-8 w-8"
              data-nimg="1"
              decoding="async"
              height="8"
              loading="lazy"
              src="/img.jpg"
              style="color: transparent;"
              width="8"
            />
          </span>
        </a>
        <div
          class="sm:hidden flex items-center space-x-3"
        >
          <button
            aria-label="Toggle mobile menu"
            class="p-2 rounded-md text-neutral-text hover:bg-background-highlight"
          >
            <svg
              class="h-6 w-6"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 6h16M4 12h16M4 18h16"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="hidden sm:flex flex-col sm:flex-row items-center mt-4 sm:mt-0 space-y-3 sm:space-y-0 sm:space-x-6"
      >
        <div
          class="relative"
        >
          <button
            class="flex items-center px-3 py-1 bg-background-elevated rounded-full text-sm text-neutral-text border border-primary-50 dark:border-primary-200 hover:bg-background-highlight transition-colors"
          >
            <div
              class="w-2 h-2 rounded-full bg-accent mr-2 animate-pulse"
            />
            <span
              class="hidden sm:inline"
              data-testid="profile-icon"
            >
              Profile
            </span>
            <svg
              class="h-4 w-4 ml-2"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 9l-7 7-7-7"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </button>
          <div
            class="absolute right-0 mt-2 w-48 bg-background-elevated rounded-md shadow-lg border border-primary-50 dark:border-primary-200 z-40"
          >
            <div
              class="p-2"
            >
              <div
                class="text-sm font-medium text-neutral-text mb-2 border-b border-primary-50 dark:border-primary-200 pb-2"
              >
                testuser@example.com
              </div>
              <button
                class="w-full text-left px-3 py-2 text-sm text-red-500 hover:bg-red-50 dark:hover:bg-red-900/20 rounded-md transition-colors"
              >
                Sign Out
              </button>
            </div>
          </div>
        </div>
        <nav
          class="flex flex-col sm:flex-row items-center space-y-2 sm:space-y-0 space-x-0 sm:space-x-2 w-full sm:w-auto"
        >
          <a
            class="nav-link w-full sm:w-auto text-center"
            href="https://www.linkedin.com/in/adam-hinton/"
            target="_blank"
          >
            Adam Hinton
          </a>
          <a
            class="nav-link w-full sm:w-auto text-center"
            href="/calculator"
          >
            Calculator
          </a>
          <a
            class="nav-link w-full sm:w-auto text-center"
            href="https://github.com/adamhinton/mileagebuddy/"
          >
            Source
          </a>
        </nav>
      </div>
    </div>
  </header>
</div>
`;
